sa/model.go

```diff
@@ -396,3 +400,181 @@ func modelToOrder(om *orderModel) (*corepb.Order, error) {
+		tokenStr = *chall.Token
+		// If the challenge status is not core.StatusPending we assume it was the 'attempted'
+		// challenge and extract the relevant fields we need.
+		if *chall.Status == string(core.StatusValid) || *chall.Status == string(core.StatusInvalid) {
+			if am.Attempted != "" || am.ValidationRecord != nil || am.ValidationError != nil {
+				return nil, errors.New("multiple challenges are non-pending")
+			}
+			am.Attempted = *chall.Type
+			// Marshal corepb.ValidationRecords to core.ValidationRecords so that we
+			// can marshal them to JSON.
+			records := make([]core.ValidationRecord, len(chall.Validationrecords))
+			for i, recordPB := range chall.Validationrecords {
+				var err error
+				records[i], err = grpc.PBToValidationRecord(recordPB)
+				if err != nil {
+					return nil, err
+				}
+			}
+			var err error
+			am.ValidationRecord, err = json.Marshal(records)
+			if err != nil {
+				return nil, err
+			}
+			// If there is a error associated with the challenge marshal it to JSON
+			// so that we can store it in the database.
+			if chall.Error != nil {
+				prob, err := grpc.PBToProblemDetails(chall.Error)
+				if err != nil {
+					return nil, err
+				}
+				am.ValidationError, err = json.Marshal(prob)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		token, err := base64.StdEncoding.DecodeString(tokenStr)
+		if err != nil {
+			return nil, err
+		}
+		am.Token = token
+	}
+
+	return am, nil
+}
+
+func modelToAuthzPB(am *authz2Model) (*corepb.Authorization, error) {
+	expires := am.Expires.UnixNano()
+	id := fmt.Sprintf("%d", am.ID)
+	v2 := true
+	pb := &corepb.Authorization{
+		V2:             &v2,
+		Id:             &id,
+		Status:         &am.Status,
+		Identifier:     &am.IdentifierValue,
+		RegistrationID: &am.RegistrationID,
+		Expires:        &expires,
+	}
+	// Populate authorization challenge array. We do this by iterating through
+	// the challenge type bitmap and creating a challenge of each type if its
+	// bit is set. Each of these challenges has the token from the authorization
+	// model and has its status set to core.StatusPending by Default. If the
+	// challenge type is equal to that in the 'attempted' row we set the status
+	// to core.StatusValid or core.StatusInvalid depending on if there is anything
+	// in ValidationError and populate the ValidationRecord and ValidationError
+	// fields.
+	challenges := int64(0)
+	for pos := uint(0); pos < 8; pos++ {
+		if (am.Challenges>>pos)&1 == 1 {
+			challType := bitToChallType[pos]
+			status := string(core.StatusPending)
+			id := challenges
+			challenges++
+			token := base64.StdEncoding.EncodeToString(am.Token)
+			challenge := &corepb.Challenge{
+				Id:     &id,
+				Type:   &challType,
+				Status: &status,
+				Token:  &token,
+			}
+			// If the challenge type is the attempted type it must be either valid
+			// or invalid and we need to populate extra fields.
+			if am.Attempted != "" && am.Attempted == challType {
```

Comment:
Can we take this whole chunk inside the "is attempted challenge" block and factor it out into its own function?